registry/client/repository.go

```diff
@@ -632,6 +635,50 @@ func (ms *manifests) Delete(ctx context.Context, dgst digest.Digest) error {
 	return HandleErrorResponse(resp)
 }
 
+func (ms *manifests) digestFromResponse(ctx context.Context, resp *http.Response, manifestURL string) (string, error) {
+	// The specs are ambivalent on Docker-Content-Digest header: on the one hand, the docs at https://docs.docker.com/registry/spec/api/#digest-header
+	// mention that "any response may include a Docker-Content-Digest header"  and on the other hand, the docs at
+	// https://docs.docker.com/registry/spec/api/#put-manifest say that a PUT request should return a Docker-Content-Digest header.
+	//
+	// Because a registry may be accessed over proxies and/or load-balancers (often happens in large corporate environments), there's no guarantee
+	// that the proxies won't prune the Docker-Content-Digest away which would result in a digest mismatch error.
+	//
+	// This function first tries to get the digest from the response headers - if that fails it falls back on getting
+	// the digest from the response body. As a last fallback, a HEAD request is made.
+	// The function may return an empty digest - it is up to the caller to determine if not receiving the digest is an error.
+	dgstHeader := resp.Header.Get("Docker-Content-Digest")
+	if dgstHeader != "" {
+		return dgstHeader, nil
+	}
+
+	// check for digest in response body if no Docker-Content-Digest header is set
+	mt := resp.Header.Get("Content-Type")
+	if mt != "" {
+		body, err := io.ReadAll(resp.Body)
+		if err == nil {
+			_, desc, err := distribution.UnmarshalManifest(mt, body)
+			if err != nil {
+				return "", err
+			}
+			dgstHeader = desc.Digest.String()
+		}
+	}
+
+	// if we don't get a digest (do we really need one?) use a HEAD request since
+	// the spec mandates this should return a Docker-Content-Digest
+	if dgstHeader == "" {
+		req, _ := http.NewRequestWithContext(ctx, http.MethodHead, manifestURL, nil)
+		r, err := ms.client.Do(req)
+		if err != nil {
+			return "", err
+		}
+
+		dgstHeader = r.Header.Get("Docker-Content-Digest")
+	}
```

Comment:
Background for the pull request: in a setting with complex network topology (and many different teams/agencies involved) against a Nexus (Sonatype) repository we first thought some load-balancers/proxies were stripping the headers. The digest was however returned as part of the body. It later turned out the repository was not sending the digest header at all in certain circumstances (when using postgreSQL as an external metadata database but would return the header when using the embedded database - this behaviour has been fixed since). So in our case it turned out to be a misbehaving repository.

The client we are using (oc-mirror) only validates the digest if it is returned - the absence of a header is not considered an error. The distribution library however returns an error when no digest response is received on PUT.  The HEAD request fallback was added as extra insurance in case the digest was not returned as part of the original headers or as part of the response body since the spec appears to mandate that a HEAD should return the digest.